libsel4vmmplatsupport/src/drivers/cross_vm_connection.c

```diff
@@ -153,57 +153,67 @@ static int reserve_event_bar(vm_t *vm, uintptr_t event_bar_address, struct conne
 
 static vm_frame_t dataport_memory_iterator(uintptr_t addr, void *cookie)
 {
-    int error;
-    cspacepath_t return_frame;
     vm_frame_t frame_result = { seL4_CapNull, seL4_NoRights, 0, 0 };
-    struct dataport_iterator_cookie *dataport_cookie = (struct dataport_iterator_cookie *)cookie;
-    seL4_CPtr *dataport_frames = dataport_cookie->dataport_frames;
-    vm_t *vm = dataport_cookie->vm;
-    uintptr_t dataport_start = dataport_cookie->dataport_start;
-    size_t dataport_size = dataport_cookie->dataport_size;
-    int page_size = seL4_PageBits;
-
-    uintptr_t frame_start = ROUND_DOWN(addr, BIT(page_size));
-    if (frame_start <  dataport_start ||
-        frame_start > dataport_start + dataport_size) {
-        ZF_LOGE("Error: Not Dataport region");
+    struct dataport_iterator_cookie *dataport = cookie;
+
+    uintptr_t frame_start = ROUND_DOWN(addr, BIT(dataport->frame_size_bits));
+    if (frame_start < dataport->base) {
+        ZF_LOGE("Access outside dataport");
+        return frame_result;
+    }
+
+    size_t offset = frame_start - dataport->base;
+    if (!IS_ALIGNED(offset, dataport->frame_size_bits)) {
```

Comment:
Maybe this should get some comments what calling contract we are expecting here, so it's more clear where the are checking violations of the calling contract vs checking bad configurations that can still happen as they are not in the scope of the calling contract. Also, we make the error messages a bit more verbose about what address exactly failed here - then debugging gets a bit easier with the addresses printed.
- It seem we expect proper alignment of `dataport->base`
- Also, `addr` is always in the range "`dataport->base` to  `dataport->num_frames * dataport->frame_size_bits`"
- We do not expect proper alignment of `addr`. But there seems there is a guarantee/expectation that the iterating loop that calls us it incrementing `addr` by what we return in `frame_result.size_bits`. So, can `addr` ever be misaligned except in the first rounds?

Reply:
Rearranged code so that `base` alignment is checked against the frame bit size. After that we have only one sanity check that `frames` inside cookie is not accessed out of bounds. But if that happens, something has gone south big time, as it seems `vm_map_reservation()` is not doing what we told it to do. Hence I feel something more heavyweight debug tool than just a print here is needed, hence I would choose not to make error message more verbose (also I couldn't fit it in 80 columns :D)

Reply:
> Maybe this should get some comments what calling contract we are expecting here, so it's more clear where the are checking violations of the calling contract vs checking bad configurations that can still happen as they are not in the scope of the calling contract. 

I realized lots of my additional checks changed the calling contract and removed those checks. My intent was to prevent anyone shooting their feet but perhaps sometimes it gets in the way. So I am now adding only the essentials.

> Also, we make the error messages a bit more verbose about what address exactly failed here - then debugging gets a bit easier with the addresses printed.

The only one added check -- an error check and not calling contract enforcement -- has now the details.

> * It seem we expect proper alignment of `dataport->base`
> * Also, `addr` is always in the range "`dataport->base` to  `dataport->num_frames * dataport->frame_size_bits`"
> * We do not expect proper alignment of `addr`. But there seems there is a guarantee/expectation that the iterating loop that calls us it incrementing `addr` by what we return in `frame_result.size_bits`. So, can `addr` ever be misaligned except in the first rounds?

These questions are now solved by mimicking the style of other mapping code, just adding enough checks that the code does not access out of bounds etc.